modules/c++/re/source/RegexPCRE.cpp

```diff
@@ -26,37 +26,148 @@
+    //       can we reuse a match data object between calls?  We then need to
+    //       handle more cases below where numMatches == 0 which indicates the
+    //       output vector wasn't large enough like the old PCRE 5.0 code used
+    //       to do.
+    ScopedMatchData(const pcre2_code* code) :
+        mCode(code),
+        mMatchData(pcre2_match_data_create_from_pattern(code, NULL))
+    {
+        if (mMatchData == NULL)
+        {
+            throw re::RegexException(Ctxt(
+                    "pcre2_match_data_create_from_pattern() failed to "
+                    "allocate memory"));
+        }
+    }
+
+    ~ScopedMatchData()
+    {
+        pcre2_match_data_free(mMatchData);
+    }
+
+    pcre2_match_data* get()
+    {
+        return mMatchData;
+    }
+
+    const PCRE2_SIZE* getOutputVector() const
+    {
+        return pcre2_get_ovector_pointer(mMatchData);
+    }
+
+    // Returns the number of matches
+    size_t match(const std::string& subject,
+                 PCRE2_SIZE startOffset = 0,
+                 sys::Uint32_T options = 0)
+    {
+        // This returns the number of matches
+        // But for no matches, it returns PCRE2_ERROR_NOMATCH
+        // Other return codes less than 0 indicate an error
+        const int returnCode =
+                pcre2_match(mCode,
+                            reinterpret_cast<PCRE2_SPTR>(subject.c_str()),
+                            subject.length(),
+                            startOffset,
+                            options,
+                            mMatchData,
+                            NULL); // Match context
+
+        if (returnCode == PCRE2_ERROR_NOMATCH)
+        {
+            return 0;
+        }
+        else if (returnCode < 0)
+        {
+            // Some error occurred
+            throw re::RegexException(Ctxt("pcre2_match() failed"));
+        }
+        else
+        {
+            return returnCode;
+        }
+    }
+
+    std::string getMatch(const std::string& str, size_t idx) const
+    {
+        const PCRE2_SIZE* const outVector = getOutputVector();
 
-re::Regex::Regex(const std::string& pattern) :
-    mPattern(pattern), mPCRE(NULL), mOvector(OVECTOR_COUNT)
+        const size_t index = outVector[idx * 2];
+        const size_t end = outVector[idx * 2 + 1];
+
+        if (end > str.length())
+        {
+            // Presumably this never happens
+            std::ostringstream ostr;
+            ostr << "Match: Match substring out of range ("
+                 << index << ", " << end << ") for string of length "
+                 << str.length();
+            throw re::RegexException(Ctxt(ostr.str()));
+        }
+
+        const size_t subStringLength = end - index;
+        return str.substr(index, subStringLength);
+    }
+
+private:
+    // Noncopyable
+    ScopedMatchData(const ScopedMatchData& );
+    ScopedMatchData& operator=(const ScopedMatchData& );
+
+private:
+    const pcre2_code* const mCode;
```

Comment:
I think `pcre2_code` is just used to get some basic info (all I know is `pcre2_match()` takes it in as const :o) )... note that `mMatchData` is only a `const` pointer, not a pointer to `const`, so that one presumably does change